### PR TITLE
`DirectoryControl`.`QueryDirectory`

### PR DESCRIPTION
**Pull request by ibeckermayer**
- Added a new method write_i8 to WriteCursor in ironrdp-pdu/src/cursor.rs, allowing for writing an 8-bit integer value to the cursor.

- Updated Rdpdr in ironrdp-rdpdr/src/lib.rs to handle an additional case for ClientDriveQueryDirectoryResponse in the error handling logic.

- Introduced several new structures in ironrdp-rdpdr/src/pdu/efs.rs related to directory information, including FileBothDirectoryInformation, FileFullDirectoryInformation, FileNamesInformation, and FileDirectoryInformation.

- Added support for the ClientDriveQueryDirectoryResponse PDU in various parts of ironrdp-rdpdr, including the RdpdrPdu enum and its implementation.

- Added a ServerDriveQueryDirectoryRequest structure to handle server requests for querying directory information, along with the necessary encoding logic.

- Added a ClientDriveQueryDirectoryResponse structure to handle client responses for directory queries, along with the necessary encoding logic.

- Updated the RDPDR PDU encoding and size calculation methods to support the new ClientDriveQueryDirectoryResponse structure.

### CODE REVIEW

**Review by CBenoit — COMMENTED** [review]
Thank you!

File: crates/ironrdp-rdpdr/src/pdu/efs.rs [hunk 1]
```diff
@@ -1931,3 +2288,103 @@ impl DeviceCloseResponse {
...
+            FileInformationClassLevel::FILE_NAMES_INFORMATION,
+        ];
+
+        if !VALID_LEVELS.contains(&file_info_class_lvl) {
+            return Err(invalid_message_err!(
+                "ServerDriveQueryDirectoryRequest::decode",
+                "file_info_class_lvl",
+                "received invalid level"
+            ));
+        }
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/pdu/efs.rs:2324]: I'm not sure that that makes sense in this case. There are [a lot of potential values](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/4718fc40-e539-4014-8e33-b675af74e3e1) for this type, of which we're only explicitly supporting a small subsection here -- these `VALID_LEVELS` plus the three additional types which are [supported by FreeRDP](https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L482) in the `ServerDriveQueryInformationRequest`. However in theory we might receive one of the many other values and/or use this same type elsewhere later down the road. I think it's best to opt for the most flexible / resilient structure for the time being.

[the remark this answers is not in the document]
Comment (CBenoit) [reply crates/ironrdp-rdpdr/src/pdu/efs.rs:2324]: Oh okay! Sounds good to me.

File: crates/ironrdp-rdpdr/src/pdu/efs.rs [hunk 2]
```diff
@@ -1931,3 +2288,103 @@ impl DeviceCloseResponse {
...
+
+        ensure_size!(in: src, size: size_of::<u8>());
+        let initial_query = src.read_u8();
+        ensure_size!(in: src, size: size_of::<u32>());
+        let path_length = cast_length!("ServerDriveQueryDirectoryRequest", "path_length", src.read_u32())?;
+        // Padding (23 bytes): An array of 23 bytes. This field is unused and MUST be ignored.
+        ensure_size!(in: src, size: 23);
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/pdu/efs.rs]: 4167fe6f56a1a4b44b955a000287c025ef26f129

File: crates/ironrdp-rdpdr/src/pdu/efs.rs [hunk 3]
```diff
@@ -1931,3 +2288,103 @@ impl DeviceCloseResponse {
...
+        let _ = src.read_slice(23);
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/pdu/efs.rs]: 4167fe6f56a1a4b44b955a000287c025ef26f129

File: crates/ironrdp-rdpdr/src/pdu/efs.rs [hunk 4]
```diff
@@ -1931,3 +2288,103 @@ impl DeviceCloseResponse {
...
+            if self.buffer.is_some() {
+                self.buffer.as_ref().unwrap().size()
+            } else {
+                0
+            }
+        )?);
+        if let Some(buffer) = &self.buffer {
+            buffer.encode(dst)?;
+        } else {
+            dst.write_u8(0) // Padding: https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L937
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/pdu/efs.rs]: 4167fe6f56a1a4b44b955a000287c025ef26f129

**Review by CBenoit — APPROVED** [review]
LGTM!

**Pull request merged**